client/allocrunner/consulsock_hook.go

```diff
@@ -0,0 +1,319 @@
+}
+
+func (h *consulSockHook) Postrun() error {
+	h.mu.Lock()
+	defer h.mu.Unlock()
+
+	if err := h.proxy.stop(); err != nil {
+		// Only log failures to stop proxies. Worst case scenario is a
+		// small goroutine leak.
+		h.logger.Debug("error stopping Consul proxy", "error", err)
+	}
+	return nil
+}
+
+type sockProxy struct {
+	allocDir *allocdir.AllocDir
+	config   *config.ConsulConfig
+
+	ctx     context.Context
+	cancel  func()
+	doneCh  chan struct{}
+	runOnce bool
+
+	logger hclog.Logger
+}
+
+func newSockProxy(logger hclog.Logger, allocDir *allocdir.AllocDir, config *config.ConsulConfig) *sockProxy {
+	ctx, cancel := context.WithCancel(context.Background())
+	return &sockProxy{
+		allocDir: allocDir,
+		config:   config,
+		ctx:      ctx,
+		cancel:   cancel,
+		doneCh:   make(chan struct{}),
+		logger:   logger,
+	}
+}
+
+// run socket proxy if allocation requires it, it isn't already running, and it
+// hasn't been told to stop.
+//
+// NOT safe for concurrent use.
+func (s *sockProxy) run(alloc *structs.Allocation) error {
+	// Only run once.
+	if s.runOnce {
+		return nil
+	}
+
+	// Only run once. Never restart.
+	select {
+	case <-s.doneCh:
+		s.logger.Trace("socket proxy already shutdown; exiting")
+		return nil
+	case <-s.ctx.Done():
+		s.logger.Trace("socket proxy already done; exiting")
+		return nil
+	default:
+	}
+
+	destAddr := s.config.GRPCAddr
+	if destAddr == "" {
+		// No GRPCAddr defined. Use Addr but replace port with the gRPC
+		// default of 8502.
+		host, _, err := net.SplitHostPort(s.config.Addr)
+		if err != nil {
+			return fmt.Errorf("error parsing Consul address %q: %v",
+				s.config.Addr, err)
+		}
+
+		destAddr = net.JoinHostPort(host, "8502")
+	}
+
+	hostGRPCSockPath := filepath.Join(s.allocDir.AllocDir, allocdir.AllocGRPCSocket)
+	listener, err := net.Listen("unix", hostGRPCSockPath)
+	if err != nil {
+		return fmt.Errorf("unable to create unix socket for Consul gRPC endpoint: %v", err)
+	}
+
+	// The gRPC socket should be usable by all users in case a task is
+	// running as an unprivileged user.  Unix does not allow setting domain
+	// socket permissions when creating the file, so we must manually call
+	// chmod afterwards.
+	// https://github.com/golang/go/issues/11822
+	if err := os.Chmod(hostGRPCSockPath, os.ModePerm); err != nil {
+		return fmt.Errorf("unable to set permissions on unix socket for Consul gRPC endpoint: %v", err)
+	}
+
+	go func() {
+		proxy(s.ctx, s.logger, destAddr, listener)
+		s.cancel()
+		close(s.doneCh)
+	}()
+
+	s.runOnce = true
+	return nil
+}
+
+// stop the proxy and blocks until the proxy has stopped. Returns an error if
+// the proxy does not exit in a timely fashion.
+func (s *sockProxy) stop() error {
+	s.cancel()
+
+	// If proxy was never run, don't wait for anything to shutdown.
+	if !s.runOnce {
+		return nil
+	}
+
+	select {
+	case <-s.doneCh:
+		return nil
+	case <-time.After(3 * time.Second):
+		return fmt.Errorf("timed out waiting for proxy to exit")
```

Comment:
This error is only logged at debug level above because if it's hit we just leak 1 goroutine per stopped allocation. It exists to assist in testing that we don't leak goroutines and never deadlock.